lib/RecommendationProvider.js

```diff
@@ -0,0 +1,157 @@
+/* globals exports, require, Task */
+"use strict";
+const {Cu} = require("chrome");
+const {setTimeout, clearTimeout} = require("sdk/timers");
+const simplePrefs = require("sdk/simple-prefs");
+
+const MILLISECONDS_IN_A_MINUTE = 1000 * 60;
+const MILLISECONDS_IN_TEN_MINUTES = MILLISECONDS_IN_A_MINUTE * 10;
+const MILLISECONDS_IN_AN_HOUR = MILLISECONDS_IN_A_MINUTE * 60;
+const MILLISECONDS_IN_A_DAY = MILLISECONDS_IN_AN_HOUR * 24;
+const MILLISECONDS_IN_A_WEEK = MILLISECONDS_IN_A_DAY * 7;
+const POCKET_PREF = "pocket.endpoint";
+
+const DEFAULT_TIMEOUTS = {
+  pocketTimeout: MILLISECONDS_IN_TEN_MINUTES, // every 10 minutes, refresh the Pocket recommendations
+};
+
+Cu.import("resource://gre/modules/Task.jsm");
+Cu.importGlobalProperties(["fetch"]);
+
+function RecommendationProvider(previewProvider, options = {}) {
+  this.options = Object.assign({}, DEFAULT_TIMEOUTS, options);
+  this._recommendedContent = [];
+  this._blockedRecommendedContent = [];
+  this._currentRecommendation = null;
+  this._pocketEndpoint = simplePrefs.prefs[POCKET_PREF];
+  this._previewProvider = previewProvider;
+  this._asyncExpireRecommendations(this.options.pocketTimeout);
+}
+
+RecommendationProvider.prototype = {
+  _pocketTimeoutID: null,
+
+  /**
+    * If we need a recommendation, go get one
+    */
+  getARecommendation(getRecommendation) {
```

Comment:
The line in question is this, right?

```javascript
let recommendation = this._recommendationProvider.getARecommendation(getRecommendation && simplePrefs.prefs.recommendations && inExperiment);
```

If your concern is added lines of code in AS, why not make this:

```javascript
let shouldGetRecommendation = getRecommendation && simplePrefs.prefs.recommendations && inExperiment;
let recommendation = shouldGetRecommendation ? this._recommendationProvider.getARecommendation() : null;
```

It might be marginally preferable to wrap it in an `if` block for readability.

It seems wasteful to do a function call when there is enough context to know that the function call was not needed. AS contains all the context needed and the boolean evaluation happens there anyway.

Another way to think about this is to think of `RecommendationProvider` as to being as decoupled from the rest of the application as possible. There is no point calling a function if no result was expected.

1. Unless I'm missing something, how does it make testing easier?
2. Point 2 is voided because AS already knows if there needs to be a recommendation or not
3. When the changes happen, we could refactor. The current code doesn't show what could be in the future yet